File: solutions/elixir/poker/2/lib/poker.ex
defmodule Poker do
  @ranks ~w(1 2 3 4 5 6 7 8 9 10 J Q K A) |> Enum.with_index() |> Map.new()

  @doc """
  Given a list of poker hands, return a list containing the highest scoring hand.

  If two or more hands tie, return the list of tied hands in the order they were received.

  The basic rules and hand rankings for Poker can be found at:

  https://en.wikipedia.org/wiki/List_of_poker_hands

  For this exercise, we'll consider the game to be using no Jokers,
  so five-of-a-kind hands will not be tested. We will also consider
  the game to be using multiple decks, so it is possible for multiple
  players to have identical cards.

  Aces can be used in low (A 2 3 4 5) or high (10 J Q K A) straights, but do not count as
  a high card in the former case.

  For example, (A 2 3 4 5) will lose to (2 3 4 5 6).

  You can also assume all inputs will be valid, and do not need to perform error checking
  when parsing card values. All hands will be a list of 5 strings, containing a number
  (or letter) for the rank, followed by the suit.

  Ranks (lowest to highest): 2 3 4 5 6 7 8 9 10 J Q K A
  Suits (order doesn't matter): C D H S

  Example hand: ~w(4S 5H 4C 5D 4H) # Full house, 5s over 4s
  """
  @spec best_hand(list(list(String.t()))) :: list(list(String.t()))
  def best_hand(hands) do
    hands
    |> Enum.map(&parse_hand/1)
    |> Enum.sort_by(&score/1, :desc)
    |> Enum.reduce([], &winner/2)
    |> Enum.map(&to_hand/1)
  end

  defp parse_hand(hand) do
    hand
    |> Enum.map(fn card ->
      [^card, rank, suit] = Regex.run(~r/(.{1,2})([CDHS])/, card)
      {rank, suit}
    end)
    |> Enum.sort_by(&Map.get(@ranks, elem(&1, 0)), :desc)
    |> maybe_transpose()
  end

  defp maybe_transpose([{"A", suit} | [{"5", _}, {"4", _}, {"3", _}, {"2", _}] = hand]) do
    hand ++ [{"1", suit}]
  end

  defp maybe_transpose(hand), do: hand

  defp to_hand(hand), do: Enum.map(hand, fn {rank, suit} -> Enum.join([ace(rank), suit]) end)

  defp ace("1"), do: "A"
  defp ace(rank), do: rank

  defp winner(hand, []), do: [hand]

  defp winner(hand, [champ | _] = champs) do
    case {score(hand), score(champ)} do
      {same, same} -> face_off([hand | champs])
      {win, loss} when win > loss -> [hand]
      _ -> champs
    end
  end

  defp score([{rank, suit}, {_, suit}, {_, suit}, {_, suit}, {_, suit}] = hand) do
    if(straight?(hand), do: 900, else: 500) + @ranks[rank]
  end

  defp score([{triplet, _}, {triplet, _}, {triplet, _}, {pair, _}, {pair, _}]) do
    600 + @ranks[triplet] * 8 + @ranks[pair]
  end

  defp score([{pair, _}, {pair, _}, {triplet, _}, {triplet, _}, {triplet, _}]) do
    600 + @ranks[triplet] * 8 + @ranks[pair]
  end

  defp score([_, {rank, _}, {rank, _}, {rank, _}, {rank, _}]), do: 700 + @ranks[rank]
  defp score([{rank, _}, {rank, _}, {rank, _}, {rank, _}, _]), do: 700 + @ranks[rank]
  defp score([{rank, _}, {rank, _}, {rank, _}, _, _]), do: 300 + @ranks[rank]
  defp score([_, {rank, _}, {rank, _}, {rank, _}, _]), do: 300 + @ranks[rank]
  defp score([_, _, {rank, _}, {rank, _}, {rank, _}]), do: 300 + @ranks[rank]

  defp score(hand) do
    grouped = Enum.group_by(hand, &elem(&1, 0))

    cond do
      Enum.count(grouped) == 3 -> 200
      Enum.count(grouped) == 4 -> 100
      straight?(hand) -> 400
      true -> 0
    end
  end

  defp straight?([{rank, _} | _] = hand) do
    rank_val = @ranks[rank]
    Enum.map(hand, &@ranks[elem(&1, 0)]) == Range.to_list(rank_val..(rank_val - 4)//-1)
  end

  defp face_off(hands) do
    case high_card(hands) do
      :tie -> hands
      :left -> [hd(hands)]
      :right -> tl(hands)
    end
  end

  defp high_card([[], []]), do: :tie
  defp high_card([_]), do: :tie
  defp high_card([[{rank, _} | left], [{rank, _} | right]]), do: high_card([left, right])

  defp high_card([[{left, _} | _], [{right, _} | _]]) do
    if @ranks[left] > @ranks[right], do: :left, else: :right
  end
end

File: solutions/clojure/poker/1/src/poker.clj
(ns poker)

(def ranks
  {"A" 0
   "K" 1
   "Q" 2
   "J" 3
   "10" 4
   "9" 5
   "8" 6
   "7" 7
   "6" 8
   "5" 9
   "4" 10
   "3" 11
   "2" 12})

(defn split-cards
  [hand]
  (map
    (fn [card]
      (let [[[c f s]] (re-seq #"(\w+)([SHCD])" card)]
        {:card c
         :face f
         :rank (ranks f)
         :suit s
         :hand hand}))
    (clojure.string/split hand #" ")))

(defn report [cards hand-rank]
  [(apply str [(conj (map :rank (sort-by :rank cards)) hand-rank)])
   (:hand (first cards))])


(defn straight-flush? [cards]
  (when (and
          (= 1 (count (set (map :suit cards))))
          (let [[a b _ _ e] (sort-by :rank cards)]
            (or (= 4 (- (:rank e) (:rank a)))
                (and (= "A" (:face a)) (= "5" (:face b))))))
    (report cards 0)))

(defn four-of-a-kind? [cards]
  (let [g (group-by :face cards)]
    (when (and (= 2 (count g))
               (let [[_ b] (sort-by (comp count second) g)]
                 (= 4 (count (second b)))))
      (report cards 1))))

(defn full-house? [cards]
  (let [g (group-by :face cards)]
    (when (and (= 2 (count g))
               (let [[a b] (sort-by (comp count second) g)]
                 (and (= 2 (count (second a)))
                      (= 3 (count (second b))))))
      (report cards 2))))

(defn flush? [cards]
  (when (= 1 (count (set (map :suit cards))))
    (report cards 3)))

(defn straight? [cards]
  (when (let [[a b c d e :as sorted] (sort-by :rank cards)]
          ;; TODO Ace low as to be sorted differently.
          (or (apply = (map + (map :rank sorted) (range 4 -1 -1)))
              (and (= "A" (:face a))
                   (= "5" (:face b))
                   (= "4" (:face c))
                   (= "3" (:face d))
                   (= "2" (:face e)))))
    (report cards 4)))

(defn three-of-a-kind? [cards]
  (let [g (group-by :face cards)]
    (when (and (= 3 (count g))
               (let [[_ _ batch] (sort-by (comp count second) g)]
                 (= 3 (count (second batch)))))
      (report cards 5))))

(defn two-pairs? [cards]
  (let [g (group-by :face cards)]
    (when (and (= 3 (count g))
               (let [[_ a b] (sort-by (comp count second) g)]
                 (and (= 2 (count (second a)))
                      (= 2 (count (second b))))))
      (report cards 6))))

(defn one-pair? [cards]
  (let [g (group-by :face cards)]
    (when (= 4 (count g))
      (report cards 7))))

(defn high-card [cards]
  (report cards 8))


(defn rank-hand [hand]
  (some identity
        (map
          #(% (split-cards hand))
          [straight-flush?
           four-of-a-kind?
           full-house?
           flush?
           straight?
           three-of-a-kind?
           two-pairs?
           one-pair?
           high-card])))


(defn best-hands [hands]
  [(second (first (sort-by first (map rank-hand hands))))])

